include/hipSYCL/sycl/buffer.hpp

```diff
@@ -607,8 +607,14 @@ class buffer : public detail::property_carrying_object
       std::copy(first, last, reinterpret_cast<T*>(&(contiguous_buffer[0])));
       copy_host_content(reinterpret_cast<T*>(contiguous_buffer.data()));
     } else {
-      std::vector<T> contiguous_buffer(num_elements);
-      std::copy(first, last, contiguous_buffer.begin());
+      std::vector<T> contiguous_buffer;
+      if constexpr (std::is_default_constructible_v<T>) {
+        contiguous_buffer.reserve(num_elements);
+        std::copy(first, last, contiguous_buffer.begin());
+      } else {
```

Comment:
Here I wasn't really sure if we need the if/else because the way it is done in the else branch works of course also for types that are default-constructible.

Reply:
Hang on, so the issue is that if it's not default constructible, we cannot reserve the memory in the vector? (does reserve also construct empty objects?) I wonder if there's a way to have a single branch using `std::copy`. Just manually allocating so that no object is initially created and then using placement new as needed?